### PR TITLE
Fixes #12154 - Limit concurrent virtual threads.

### LINKED ISSUE #12154: Is it possible to define max number of virtual threads when VirtualThreadsExecutor is enabled, i.e. max number of http requests being handled in the same time?
**12.0.12**

**Java 21**

**Question**
Hi!
Is it possible to define max number of virtual threads when VirtualThreadsExecutor is enabled, i.e. max number of http requests being handled in the same time?

I tried to dig into source code and as a result assumed that _any_ request submitted to Jetty is dispatched to virtual thread executor via AdaptiveExecutionStrategy as is so it doesn't mater 1k or 1kk requests came. 
Thanks!

### PR DESCRIPTION
**Pull request by sbordet**
Updated VirtualThreadPool to limit the number of concurrent virtual threads using a Semaphore.

Updated modules and documentation.

### CODE REVIEW

**Comment by gregw** [comment]
@sbordet can we make this do if anybody configures the limit to less than 10000 it does "warning: just use normal threads. The emperor is indeed naked!"

**Comment by sbordet** [comment]
@gregw sure, but I think you meant "less than a million" 🤣

File: jetty-core/jetty-server/src/main/config/modules/threadpool-all-virtual.mod [hunk 1]
```diff
@@ -13,7 +13,16 @@ etc/jetty-threadpool-all-virtual.xml
 
 [ini-template]
 # tag::documentation[]
-## Platform threads name prefix.
+## Virtual threads name prefix.
 #jetty.threadPool.namePrefix=vtp<hashCode>
 
+## Maximum number of concurrent virtual threads.
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/config/modules/threadpool-all-virtual.mod]: I don't think "concurrent" is correct or necessary here.  It makes it sound like there could be a lot more virtual threads, but only a limited number concurrently executing (which is a factor of cores and/or native threads).
```suggestion
## Maximum number of live virtual threads.
```

File: jetty-core/jetty-server/src/main/config/modules/threadpool-all-virtual.mod [hunk 2]
```diff
@@ -13,7 +13,16 @@ etc/jetty-threadpool-all-virtual.xml
...
+#jetty.threadPool.maxThreads=200
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/config/modules/threadpool-all-virtual.mod:20]: 200?!?!?!?  aren't you missing a few orders of magnitude here?   Nobody should be using virtual threads (and risking all sorts of bad behaviour) at 200 threads... maybe 2,000, but better 20,000
```suggestion
#jetty.threadPool.maxThreads=20000
```

File: jetty-core/jetty-server/src/main/config/modules/threadpool-virtual.mod [hunk 3]
```diff
@@ -37,6 +37,10 @@ etc/jetty-threadpool-virtual.xml
 ## Virtual threads name prefix.
 #jetty.threadPool.virtual.namePrefix=qtp<hashCode>-virtual-
 
-## Whether virtual threads inherits the values of inheritable thread locals.
-#jetty.threadPool.virtual.inheritInheritableThreadLocals=true
+## Max number of concurrent virtual threads.
+#jetty.threadPool.virtual.maxThreads=200
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/config/modules/threadpool-virtual.mod:41]:
```suggestion
#jetty.threadPool.virtual.maxThreads=20000
```

Comment (gregw) [reply jetty-core/jetty-server/src/main/config/modules/threadpool-virtual.mod:41]: Why is this `jetty.threadPool.virtual.maxThreads` here and `jetty.threadPool.maxThreads` in `threadpool-all-virtual.mod`

**Review by gregw — CHANGES_REQUESTED** [review]

**Review by gregw — DISMISSED** [review]

**Pull request merged**